### PR TITLE
Storage Based Retention

### PR DESCRIPTION
**Pull request by NickM-27**
To - Do:

- [x] Handle case where only retained segments are left and they need to be deleted
- [x] Likely need to add some tests around the cleanup and other logic
- [x] Update docs to reflect change in behavior
- [x] Handle edge case of cameras that only record sometimes (ex: a camera that only records when the garage is open)

### CODE REVIEW

**Comment by NickM-27** [comment]
I am unsure if when running `reduce_storage_consumption` a camera's current record status (enabled / disabled) should be taken into account. I think my current logic covers that and it shouldn't matter, but curious what others think.

**Review by blakeblackshear — CHANGES_REQUESTED** [review]
Good first pass. I just think we can get there in a way that's less complex and more efficient.

File: frigate/storage.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+        self.avg_segment_sizes["total"] = {
+            "segment": total_avg_segment,
+            "hour": total_avg_hour,
+        }
+
+    def check_storage_needs_cleanup(self) -> bool:
+        """Return if storage needs cleanup."""
+        # currently runs cleanup if less than 1 hour of space is left
+        remaining_storage = round(shutil.disk_usage(RECORD_DIR).free / 1000000, 1)
```
[the remark this answers is not in the document]
Comment (NickM-27) [reply frigate/storage.py:76]: My unraid setup has a write cache which is where frigate recordings go until the cache gets to a threshold then the recordings are moved to a hard drive until frigate deletes them.

There may be a better way to test this, but I spun all my drives down on Unraid. Frigate recordings were on and being written to the cache drive. I got a shell in frigate and ran this command in python and the sizes were going up over time, but the drive did not spin up. 

I then ran `ls /media/frigate/` and the drive spun up. So we should be good there.

File: frigate/storage.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+                7200 / self.avg_segment_sizes[camera]["segment_duration"]
+            )
+            recordings: Recordings = (
+                Recordings.select()
+                .where(Recordings.camera == camera)
+                .order_by(Recordings.start_time.asc())
+                .limit(segment_count * 12)
+            )
+
+            # cameras that are only recording part time
```
[the remark this answers is not in the document]
Comment (NickM-27) [reply frigate/storage.py:145]: Right, this was mainly due to the fact that I am currently removing recordings from each camera which meant a camera that only recorded occasionally would be required to delete all of its retained segments. With the approach you described above this won't be necessary.

**Review by blakeblackshear — APPROVED** [review]

File: frigate/storage.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                deleted_segments_size += recording.segment_size
+                Path(recording.path).unlink(missing_ok=True)
+                deleted_recordings.add(recording.id)
+
+        # need to delete retained segments
+        if deleted_segments_size < hourly_bandwidth:
+            logger.error(
+                f"Could not clear {hourly_bandwidth} currently {deleted_segments_size}, retained recordings must be deleted."
+            )
+            for recording in recordings.objects().iterator():
```
[the remark this answers is not in the document]
Comment (NickM-27) [reply frigate/storage.py:140]: The test `def test_storage_cleanup(self):` does capture this so I think it does work this way, but I'll explicitly define it just to be sure

**Comment by NPutting** [comment]
@NickM-27 Should it be pretty easy to add a config of size to keep free instead of the 1 hour calculation? I think the hour cleanup at a time would still work but having a config for that might be nice as well. Wanted to get your thoughts before I tried to implement it.

**Comment by NickM-27** [comment]
> @NickM-27 Should it be pretty easy to add a config of size to keep free instead of the 1 hour calculation? I think the hour cleanup at a time would still work but having a config for that might be nice as well. Wanted to get your thoughts before I tried to implement it. 

It was deliberately left at this for now as our focus for 0.12.0 is stability. In a future release more config options are planned. 

That is to say: feel free to go for it but not sure it'll be included in the upcoming release. That being said, it should be fairly trivial to implement.

**Comment by NPutting** [comment]
> > @NickM-27 Should it be pretty easy to add a config of size to keep free instead of the 1 hour calculation? I think the hour cleanup at a time would still work but having a config for that might be nice as well. Wanted to get your thoughts before I tried to implement it.
> 
> It was deliberately left at this for now as our focus for 0.12.0 is stability. In a future release more config options are planned.
> 
> That is to say: feel free to go for it but not sure it'll be included in the upcoming release. That being said, it should be fairly trivial to implement.

Awesome, Thanks for the quick reply. It makes sense that this initial feature falls under stability improvements. I'll try and make a draft PR with a configurable limit option and understand if it's pushed off on merging for awhile..

**Comment by XtremeOwnageDotCom** [comment]
@blakeblackshear 

Make sure to tag/close [#994](https://github.com/blakeblackshear/frigate/issues/994) with this PR.

(I can't touch it. But, this PR resolves that issue)

**Comment by NickM-27** [comment]
@XtremeOwnageDotCom I don't think that issue can be closed, there's more configuration and other things that need to be done to fully satisfy that request.

**Comment by XtremeOwnageDotCom** [comment]
@NickM-27  It meets the requirement for which I originally opened that issue for.

However, I gave up trying to keep up with all of the riff-raff which occurred down in the issue when it went off the rails..
